Fixing invalidate_all

lfvilella opened the pull request:
I was trying to use the `invalidate_all` then I realize it was not working, it was a problem in the `cache_key` been generated in two different ways, so follow the PR to:

1) fix the problem on the `invalidate_all`
2) refactoring the tests to be more readable
3) adding a CI to test the code in all supported python versions
4) update the readme adding some links to it

taylorhakes commented:
@lfvilella Thank you for the PR. Can you explain the issue you are fixing a little? Why does it need to switch from using `zrange` to `keys`. `keys` is not meant to be used in production. See the warning here: https://redis.io/commands/keys

lfvilella commented:
Hi @taylorhakes well pointed, I should've added a better explanation on the PR description, but basically the problem I'm facing is, the `invalidate_all` is not invalidating the cache, checkout the example bellow.

```py
import uuid

import redis
import redis_cache

client = redis.StrictRedis(host="redis-test-host", decode_responses=True)
cache = redis_cache.RedisCache(redis_client=client)

@cache.cache()
def func_1(*args, **kwargs):
    return str(uuid.uuid4())

r1 = func_1(1, 2)
r2 = func_1(1, 2)

func_1.invalidate_all()

r3 = func_1(1, 2)

assert r1 == r2 != r3
```

```py
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AssertionError
```

Related to `zrange` vs `keys` thanks for pointing it out, I didn't realize it wasn't production ready, I'll try to move it back to `zrange` and I'll ping you again.

lfvilella commented:
@taylorhakes I reverted back to `zrange` and added `zadd` to make it works, did some tests seems to be working fine, if could add your thoughts to that.

taylorhakes commented:
@lfvilella Thank you for providing that test case. I now see the issue you were talking about. I think zadd is not the best way to fix this issue. It has a few problems.

1. It stores the keys twice (more memory) for the only purpose of `invalidate_all`. It's a function that isn't used that often.
2. If the key has a `ttl` and no `limit` we will keep storing keys forever even after they expire.
3. It adds an extra round trip to the redis server, slowing down the caching

My suggestion is to just change `invalidate_all` instead. We can use code similar to your original change. See below.

```python
# Utility function to batch keys
def batcher(iterable, n):
    args = [iter(iterable)] * n
    return izip_longest(*args)

def invalidate_all(self, *args, **kwargs):
    for keys in batcher(self.client.scan_iter(f'{self.prefix}:{self.namespace}:*'), 500):
        self.client.delete(*keys)
```

This is has the downside of making `invalidate_all` slower, but fixes the issues I mentioned above. What do you think?

Thanks for helping with the other changes. I am OK with them, but please in the future create an issue before making any major changes. I will add a note on this repo, but it's a good practice to do on all repos. Often maintainers won't be as receptive to major changes to their repo without discussion. It will save you a lot of wasted work.

Could you change the redis version from 6 to 5? I want to make sure we keep compatibility with 5, which is the lowest supported version currently.

lfvilella commented:
@taylorhakes your suggestion worked fine, it's a known fact that cache invalidation is hard, and as you said `invalidate_all` is not that used, so imo having it more slow it's acceptable, we could add some optimization  flags to next version then developers could choose having it faster but costing more memory.

About creating an issue before sending changes you are totally right, I was playing around with your lib then get excited and started making changes, for sure is not the way to go, I'll make sure I create an issue for the next ones.

Now talking about the redis version I added tests for redis 5, redis 6 and also added a trigger for `latest` tags so when a new version of redis goes live we will know how the code is performing on it.

taylorhakes commented on redis_cache/__init__.py:
```diff
@@ -119,6 +138,9 @@ def inner(*args, **kwargs):
                 get_cache_lua_fn(self.client)(keys=[key, self.keys_key], args=[result_serialized, self.ttl, self.limit])
+
+                score = self.client.zcount(self.keys_key, 0, -1)
+                self.client.zadd(self.keys_key, {key: score})
```
Is this necessary anymore?

lfvilella replied:
Oops good catch forgot to delete that.

taylorhakes commented:
Thanks @lfvilella for all your hard work on this. I really appreciate it.
